engine/api/events.go

```diff
@@ -0,0 +1,217 @@
+		if c.User.Username == username {
+			c.User = user
+		}
+	}
+	b.mutex.Unlock()
+}
+
+// Start the broker
+func (b *eventsBroker) Start(c context.Context) {
+	for {
+		select {
+		case <-c.Done():
+			// Close all channels
+			b.mutex.Lock()
+			for c := range b.clients {
+				delete(b.clients, c)
+			}
+			b.mutex.Unlock()
+			if c.Err() != nil {
+				log.Error("eventsBroker.CacheSubscribe> Exiting: %v", c.Err())
+				return
+			}
+		case receivedEvent := <-b.messages:
+			bEvent, err := json.Marshal(receivedEvent)
+			if err != nil {
+				continue
+			}
+
+			b.mutex.Lock()
+			for _, i := range b.clients {
+				if i.User.Admin {
+					i.Queue <- string(bEvent)
+					continue
+				}
+				manageEvent(receivedEvent, string(bEvent), i)
+			}
+
+			b.mutex.Unlock()
+		}
+	}
+}
+
+func (b *eventsBroker) ServeHTTP() Handler {
+	return func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+		// Make sure that the writer supports flushing.
+		f, ok := w.(http.Flusher)
+		if !ok {
+			http.Error(w, "Streaming unsupported!", http.StatusInternalServerError)
+			return nil
+		}
+
+		uuid, errS := sessionstore.NewSessionKey()
+		if errS != nil {
+			return sdk.WrapError(errS, "eventsBroker.Serve> Cannot generate UUID")
+		}
+
+		user := getUser(ctx)
+		if err := loadUserPermissions(b.dbFunc(), b.cache, user); err != nil {
+			return sdk.WrapError(err, "eventsBroker.Serve Cannot load user permission")
+		}
+
+		messageChan := eventsBrokerSubscribe{
+			UIID:   string(uuid),
+			User:   user,
+			Events: make(map[int]bool),
+			Queue:  make(chan string, 10), // chan buffered, to avoid goroutine Start() wait on push in queue
+		}
+		messageChan.Events[EventLastUpdate] = true
+
+		// Add this client to the map of those that should receive updates
+		b.mutex.Lock()
+		b.clients[string(uuid)] = messageChan
+		b.mutex.Unlock()
+
+		// Set the headers related to event streaming.
+		w.Header().Set("Content-Type", "text/event-stream")
+		w.Header().Set("Cache-Control", "no-cache")
+		w.Header().Set("Connection", "keep-alive")
+		w.Header().Set("X-Accel-Buffering", "no")
+
+		fmt.Fprint(w, "data: ACK\n\n")
+		f.Flush()
+
+		tick := time.NewTicker(time.Second)
+	leave:
+		for {
+			select {
+			case <-ctx.Done():
+				b.mutex.Lock()
+				delete(b.clients, messageChan.UIID)
+				b.mutex.Unlock()
+				break leave
+			case <-w.(http.CloseNotifier).CloseNotify():
+				b.mutex.Lock()
+				delete(b.clients, messageChan.UIID)
+				b.mutex.Unlock()
+				break leave
+			case msg := <-messageChan.Queue:
+				w.Write([]byte("data: "))
+				w.Write([]byte(msg))
+				w.Write([]byte("\n\n"))
+				f.Flush()
+			case <-tick.C:
+				f.Flush()
+			}
+		}
+		return nil
+	}
+}
+
+func manageEvent(event sdk.Event, eventS string, subscriber eventsBrokerSubscribe) {
+	if strings.HasPrefix(event.EventType, "EventProject") {
+		if permission.ProjectPermission(event.ProjectKey, subscriber.User) >= permission.PermissionRead {
+			subscriber.Queue <- eventS
+			return
```

Comment:
I think you can put return just below inside the strings.HastPrefix condition. In order to not test all other if strings.HasPrefix